### ISSUE TITLE
Implement URI encoding and decoding for filenames

### ISSUE THREAD

**Issue by andy5995** [theimpossibleastronaut/rmwrs#9]
When the absolute path and filename is written to a .trashinfo file, the string should be encoded per [RFC2396](https://tools.ietf.org/html/rfc2396).

Conversely, the string will need to be decoded when restoring the file (#5).

This already happens for the C version of rmw. Here are the relevant functions that it uses:

`is_unreserved()`
https://github.com/theimpossibleastronaut/rmw/blob/94d19a0017b4d749a03baf6277f4f983a1dcfb3a/src/utils_rmw.c#L190-L220

`escape_url()`
https://github.com/theimpossibleastronaut/rmw/blob/94d19a0017b4d749a03baf6277f4f983a1dcfb3a/src/utils_rmw.c#L223-L280

`unescape_url()`
https://github.com/theimpossibleastronaut/rmw/blob/94d19a0017b4d749a03baf6277f4f983a1dcfb3a/src/utils_rmw.c#L283-L336

**Comment by andy5995** [comment]
I just found there's a library developed by @johnstonskj that may work for this ticket.

[rdftk_iri](https://lib.rs/crates/rdftk_iri)

**Comment by andy5995** [comment]
I've applied [a patch](https://github.com/theimpossibleastronaut/rmwrs/commit/10cd9a03dea53777d95a7e9de813a4679ecfdc6b) that uses the [pct-str](https://lib.rs/crates/pct-str) library. It follows the [RFC 3986](https://tools.ietf.org/html/rfc3986). It shows that [2396](https://tools.ietf.org/html/rfc2396) has been deprecated.

The [Freedesktop Trash spec](https://specifications.freedesktop.org/trash-spec/trashspec-latest.html) still shows RFC 2396 as the required spec for path encoding, but I'm guessing that info will be updated.

One difference with the 3986 spec is that slashes are escaped now.

Here's the resulting Trash Info file with the patch applied, after I run `rmwrs 'foo bar!'`

```
[Trash Info]
Path=%2Fhome%2Fandy%2Fsrc%2Ftheimpossibleastronaut%2Frmwrs%2Ffoo bar%21
DeletionDate=2020-09-15T14:25:14
```

**Comment by andy5995** [comment]
I think best to stick with the Freedesktop trash spec, using 2396. We can still use the pct_str library; there's an example here for customizing the reserved and unreserved characters:

https://docs.rs/pct-str/1.1.0/pct_str/

**Comment by andy5995** [comment]
I think best to forget about using a library and just convert to Rust the three functions mentioned in the first post.

**Comment by andy5995** [comment]
Resolved in #10 

Thank you @Jammyjamjamman !
